### PR TITLE
Implement Modules

### PR DESCRIPTION
**Pull request by Shrinjay**
## Implementation Description
This PR involves adding modules to the database as a layer between courses and lessons, modules are added as a property of course objects. 

## What should reviewers focus on?
- Ensure that my validation logic is valid, I built a custom object to validate objects in the future which might come in handy
- Ensure that the serialization stuff won't break
- Ensure that the modules data model and all is implemented correctly

## Testing Procedure
- Try passing any invalid inputs 

## Checklist
- [ ] Ensure code follows style guide by running the linters
- [ ] I have updated the documentation or deemed it unnecessary
- [ ] I have linked the relevant issue in this PR
- [ ] I have requested a review from the PL, as well as other dev(s) (and designers if necessary)

### CODE REVIEW

File: backend/package.json
Comment (draymond63) [inline backend/package.json:32]: This is no longer true right? Could you remove this from the commit?

File: backend/services/interfaces/ICourseService.ts
Comment (draymond63) [inline backend/services/interfaces/ICourseService.ts:62]: I wonder if we could extend/partial the regular DTOs, and simply replace the modules attribute with the correct type. Is that something we could do? It would definitely cut down on the duplication of code.

File: backend/graphql/resolvers/courseResolvers.ts
Comment (draymond63) [inline backend/graphql/resolvers/courseResolvers.ts:113]: Some small linter warnings, could you either resolve them or disable them?

**Review by draymond63 — CHANGES_REQUESTED** [review]
Good work!

The logic is all there and I'm happy with the implementation. Just some nitpicks here and there :)

File: backend/middlewares/validators/courseValidators.ts
Comment (Shrinjay) [reply backend/middlewares/validators/courseValidators.ts:16]: It is, the validator should be able to detect that it's an array and cast this over an array but I realize this is a bit messy so I'll find a better way to mark something as an array.

File: backend/services/interfaces/ICourseService.ts
Comment (Shrinjay) [reply backend/services/interfaces/ICourseService.ts:62]: We definitely could, I don't really know why I did this tbh

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:37]: Think we should make `{ [id: string]: Module }` its own type, it's used a decent amount and is something that should be very clear

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:50]: nitpick: ehh I think this is fairly self-explanatory, don't know if the comment is really that helpful

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:51]: When do we expect the `moduleDTO` to have an ID vs. not have an ID? If it should always have an ID, we should expect to always have an ID and handle this more forcefully rather than absorbing the error (whether this is showing an error, or making a mutation in the db to fix)

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:69]: - Doesn't `course` always exist? Why the `?` operator

Would this work?
```suggestion
  return {
    ...course,
    modules: course.modules ? serializeToModuleDTO(course.modules) : [],
  };
```

Comment (e-wai) [reply backend/graphql/resolvers/courseResolvers.ts:69]: Also, shouldn't `modules` always exist? It might just be empty (I don't remember how it's implemented right now, but initially the `lessons` parameter was required on create course)

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:85]: same comments as on `serialize` (less verbose version, unexpected usages of `?`)

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [inline backend/graphql/resolvers/courseResolvers.ts:152]: In terms of naming, are we expecting a `serializedCourse` response? I don't think so, because we would serialize it when it's returned -- think the param name on ln#148 doesn't make sense then

nitpick syntax again, does this work?

```suggestion
        .then((serializedCourse) => serializedCourse ? seralizeCourseResponse(serailizedCourse) : null;
```

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:11]: If this is a return type, shouldn't they all be non-optional fields? Doesn't mean the user has to ask for them but they have to all exist

nitpick: for consistency, `ModuleResponseDTO`

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:21]: Think lessons should always exist, but it can be empty. Also, should always be of type string:
```suggestion
    lessons: [String!]!
```

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:40]: Also think these should be `modules: [ModuleDTO!]!`

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:33]: same comment as on `ModuleDTO`, if it's a response I think they should always be required

File: backend/graphql/types/courseType.ts
Comment (e-wai) [reply backend/graphql/types/courseType.ts:21]: also, nitpick: `ModuleInputDTO` -> `ModuleRequestDTO`

**Review by e-wai — CHANGES_REQUESTED** [review]
I'll split the review into two, this is the first half (got somewhat nitpicky, sorry)

File: backend/middlewares/validators/util.ts
Comment (e-wai) [inline backend/middlewares/validators/util.ts:48]: If I'm remembering my falsy values right, this can all become `return (object) && ...`

File: backend/middlewares/validators/util.ts
Comment (e-wai) [inline backend/middlewares/validators/util.ts:53]: Bit unclear on this piece of logic, what's happening here?

File: backend/graphql/resolvers/courseResolvers.ts
Comment (Shrinjay) [reply backend/graphql/resolvers/courseResolvers.ts:37]: Yeahh I should've done that, just got lazy, I'll make it its own type

File: backend/graphql/resolvers/courseResolvers.ts
Comment (Shrinjay) [reply backend/graphql/resolvers/courseResolvers.ts:50]: It kind of breaks the pattern of generating IDs via MongoDB, but this needs to happen since the modules don't have their own objectID like courses do as they're not in their own collection, that's why I made this a comment.

File: backend/graphql/resolvers/courseResolvers.ts
Comment (Shrinjay) [reply backend/graphql/resolvers/courseResolvers.ts:51]: Module DTO might not have an ID when we have a request to create a new course with a new module, in which case we create the ID here, but I agree that doing this more forcefully elsewhere might be cleaner. The other option is generating an ID in the client but that's even more anti-pattern.

File: backend/graphql/resolvers/courseResolvers.ts
Comment (Shrinjay) [reply backend/graphql/resolvers/courseResolvers.ts:69]: I wasn't sure, iirc lessons was optional in the graphql type previously so I made modules optional as well, and I think it makes sense to have a newly created course with no modules. Although now that I think about it, it makes more sense to remove this handling and have the frontend always send `[]` to indicate no modules.

File: backend/graphql/resolvers/courseResolvers.ts
Comment (Shrinjay) [reply backend/graphql/resolvers/courseResolvers.ts:152]: Yeah good catch, I just messed up the naming

File: backend/graphql/types/courseType.ts
Comment (Shrinjay) [reply backend/graphql/types/courseType.ts:11]: I made them optional because the pattern seemed to be that most fields were left optional, didn't really think through it, but yes making them non-optional makes more sense

File: backend/middlewares/validators/util.ts
Comment (Shrinjay) [reply backend/middlewares/validators/util.ts:53]: Basically, `string[]` represents an array, so if the last 2 chars of a string are '[]', then we should use validate array. When passing into validateArray, we only want the type of each item, so we remove the `[]` characters.

File: backend/graphql/resolvers/courseResolvers.ts
Comment (e-wai) [reply backend/graphql/resolvers/courseResolvers.ts:50]: OK fair enough

File: backend/graphql/resolvers/courseResolvers.ts
Comment (y2kwak) [inline backend/graphql/resolvers/courseResolvers.ts:31]: nit: probably doesn't affect logic; but consider using const arrow functions to keep it consistent w the rest of the file

File: frontend/src/APIClients/CourseAPIClient.ts
Comment (y2kwak) [inline frontend/src/APIClients/CourseAPIClient.ts:143]: I might not be following correctly, but do we need to deserialize the course response here?

**Review by y2kwak — COMMENTED** [review]
this seemed like a complicated problem ngl, but your solution makes sense - good job :))

File: backend/middlewares/validators/util.ts
Comment (e-wai) [reply backend/middlewares/validators/util.ts:53]: Makes sense, maybe worth a comment

File: frontend/src/APIClients/CourseAPIClient.ts
Comment (Shrinjay) [reply frontend/src/APIClients/CourseAPIClient.ts:143]: Yes we do, good catch!

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:30]: Why not `[ModuleResponseDTO!]!` -- ie. why no strict type check on `ModuleResponseDTO`, do we not expect that the response will always be of type `ModuleResponseDTO`? If it's because of missing ID, I'd rather we make id optional on the other DTO object

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:40]: Same Q as above

File: backend/graphql/types/courseType.ts
Comment (e-wai) [inline backend/graphql/types/courseType.ts:50]: Same ^

File: frontend/src/APIClients/CourseAPIClient.ts
Comment (e-wai) [inline frontend/src/APIClients/CourseAPIClient.ts]: nitpick: would it make sense to call this `ModulesById`? It makes the usage much more obvious

**Review by e-wai — APPROVED** [review]
Looks great 🥳 , some tiny comments left but nothing blocking

- I didn't get around to testing this manually (but I can if no one else has checked it out)
- Could you add the link to the data models updated page in the PR description and add some context about why the serialization is necessary, other non-obvious details about when we expect id, etc.?
- Maybe just check with @y2kwak to see if she's waiting for this to merge before/after the GraphQL stuff? (I think after, but maybe just clear that up to avoid merge conflicts)

**Review by draymond63 — APPROVED** [review]
Things look good to me, although Emma did have some good comments about enforcing the types within arrays. There's also a few linter things, but I'm good with this PR getting merged to prevent other things from getting blocked!

File: frontend/src/APIClients/CourseAPIClient.ts
Comment (Shrinjay) [reply frontend/src/APIClients/CourseAPIClient.ts]: That works too

File: backend/graphql/types/courseType.ts
Comment (Shrinjay) [reply backend/graphql/types/courseType.ts:30]: Honestly I just didn't know that you needed to do [ModuleResponseDTO!]!

**Comment by Shrinjay** [comment]
Just went ahead and merged this after it built, I fixed what Emma talked about

**Pull request merged**
